From patchwork Sun Jan  3 11:06:32 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Samuel Holland <samuel@sholland.org>
X-Patchwork-Id: 11995647
Return-Path: 
 <SRS0=N8P+=GG=lists.infradead.org=linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-17.0 required=3.0 tests=BAYES_00,DKIMWL_WL_HIGH,
	DKIM_SIGNED,DKIM_VALID,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,
	INCLUDES_PATCH,MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,USER_AGENT_GIT
	autolearn=ham autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 91A2BC433E0
	for <linux-arm-kernel@archiver.kernel.org>;
 Sun,  3 Jan 2021 11:08:46 +0000 (UTC)
Received: from merlin.infradead.org (merlin.infradead.org [205.233.59.134])
	(using TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by mail.kernel.org (Postfix) with ESMTPS id 535E820771
	for <linux-arm-kernel@archiver.kernel.org>;
 Sun,  3 Jan 2021 11:08:46 +0000 (UTC)
DMARC-Filter: OpenDMARC Filter v1.3.2 mail.kernel.org 535E820771
Authentication-Results: mail.kernel.org;
 dmarc=none (p=none dis=none) header.from=sholland.org
Authentication-Results: mail.kernel.org;
 spf=none
 smtp.mailfrom=linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@lists.infradead.org
DKIM-Signature: v=1; a=rsa-sha256; q=dns/txt; c=relaxed/relaxed;
	d=lists.infradead.org; s=merlin.20170209; h=Sender:Content-Transfer-Encoding:
	Content-Type:Cc:List-Subscribe:List-Help:List-Post:List-Archive:
	List-Unsubscribe:List-Id:MIME-Version:References:In-Reply-To:Message-Id:Date:
	Subject:To:From:Reply-To:Content-ID:Content-Description:Resent-Date:
	Resent-From:Resent-Sender:Resent-To:Resent-Cc:Resent-Message-ID:List-Owner;
	 bh=xODDDfMYjGP/If1s+77LQF3utnLKDtSEx64F57nsu5o=; b=u41kUXn2RZ0usncGqY3Xy8OOF
	RVskaj3/Ahn3uOtel+Q9vHiNRX4DyfMpbEd0VHiX0BDoHhICKqh8QLXYUKsRqZ1yVqAo2SU5ITwmj
	d1tcsd6yWnZQ8Tatvjq7S2m5Dey2It9X4FutiDTm7Vc/mQUDegfz6HxH/06ECs2GBvt8icBEuxbcl
	FEm1P81KLQMhRW4KzNF7OMc6JB8Svusa5jdo3UraX8JdRiAfftzxS+NaqM3F2m38fmyvxCGeogMXC
	aWWuROsCQkdX47k+oFKDhq25w3x2AEi95+ccQspECdMDB0rKHSoBUzfl7zKOpxRqgWmzDvc/zeBxA
	v73YcZVvw==;
Received: from localhost ([::1] helo=merlin.infradead.org)
	by merlin.infradead.org with esmtp (Exim 4.92.3 #3 (Red Hat Linux))
	id 1kw1Do-0002RI-88; Sun, 03 Jan 2021 11:06:52 +0000
Received: from wout5-smtp.messagingengine.com ([64.147.123.21])
 by merlin.infradead.org with esmtps (Exim 4.92.3 #3 (Red Hat Linux))
 id 1kw1Dh-0002No-Ja
 for linux-arm-kernel@lists.infradead.org; Sun, 03 Jan 2021 11:06:48 +0000
Received: from compute5.internal (compute5.nyi.internal [10.202.2.45])
 by mailout.west.internal (Postfix) with ESMTP id 0919D544;
 Sun,  3 Jan 2021 06:06:38 -0500 (EST)
Received: from mailfrontend1 ([10.202.2.162])
 by compute5.internal (MEProxy); Sun, 03 Jan 2021 06:06:39 -0500
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=sholland.org; h=
 from:to:cc:subject:date:message-id:in-reply-to:references
 :mime-version:content-transfer-encoding; s=fm1; bh=WyR8h/jVWSn2P
 IgW0OwRkxU8VaseOEGPaTS0rZRolAU=; b=BIP7jcuLEYLe8CBmY2ptXwkUK/50j
 TzKCZ6IdcQ/lFqQwZ7n/WldWxe6ZObEzwttaz6oiIYmoeVqmtSacYw1+ZKdiwbH9
 sslCTDGLYpaHpHtgFL8H3KMh+3OhLyFiACwdDPjicTlDINppdRoLw831+3GefODK
 YLYnW9F1Tse9rj0X/9/hmJsSg7KTO30ow6b+0FqBmA+DxjdHufsaw9ze/22I82o7
 gvHXG79Sji4rvYNtaDGfNZLuVh84dy0DqnzoVfPZXHtMDqRZ2eB7uRGzy02ndfp/
 JtpYmaun/Tw7513mjgtWdjvq7j13UKpdA/AGaZd3RaGVU0aajL+1ciduA==
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=
 messagingengine.com; h=cc:content-transfer-encoding:date:from
 :in-reply-to:message-id:mime-version:references:subject:to
 :x-me-proxy:x-me-proxy:x-me-sender:x-me-sender:x-sasl-enc; s=
 fm1; bh=WyR8h/jVWSn2PIgW0OwRkxU8VaseOEGPaTS0rZRolAU=; b=hQdWEOVc
 WBhfNNYc7D/JxZ5eqvsCLDCup+71pRAALbuEz9nR2MHBhiV8R7ySSfkQE+1SWuXV
 8YX6QSg3H0nzzVfFzIp5giZUVwX9FzQ9i9VbdgMJWbm57oP6jqs5KdDf1DeW1f63
 eT1TjnPM3RIpA+avqcIYSJc6l1FhJTuImiZzSq6qX3Gt0qvy2xBtbOEOMs9w20cK
 BTJwuhT37E1R9Mdf6g7ezBbILrqtJJiU/GUQLawGvmI7dpUeeCaj2sk/32k/twht
 VVwpLJRq2N6l09rCeh1guUf4ebSQRTqTsLj7FLwZo5rhj5+ivlRctiHAyRkRRvur
 KRqVzlx8ZfCJig==
X-ME-Sender: <xms:PKXxX4XxwM01G2mGF9y-PN9K0PityJGJi-fjCWjFbbq4Azn1Mm3XTA>
 <xme:PKXxX8lvtjjwwfYX872VicrY4JKpCnG0FOxtuFMvaK-JP016bLra3qVFCk2NArPbJ
 Aj77WJEjgj8ouXrbw>
X-ME-Proxy-Cause: 
 gggruggvucftvghtrhhoucdtuddrgedujedrvdefuddgvdegucetufdoteggodetrfdotf
 fvucfrrhhofhhilhgvmecuhfgrshhtofgrihhlpdfqfgfvpdfurfetoffkrfgpnffqhgen
 uceurghilhhouhhtmecufedttdenucesvcftvggtihhpihgvnhhtshculddquddttddmne
 cujfgurhephffvufffkffojghfggfgsedtkeertdertddtnecuhfhrohhmpefurghmuhgv
 lhcujfholhhlrghnugcuoehsrghmuhgvlhesshhhohhllhgrnhgurdhorhhgqeenucggtf
 frrghtthgvrhhnpeduhfejfedvhffgfeehtefghfeiiefgfeehgfdvvdevfeegjeehjedv
 gfejheeuieenucfkphepjedtrddufeehrddugeekrdduhedunecuvehluhhsthgvrhfuih
 iivgeptdenucfrrghrrghmpehmrghilhhfrhhomhepshgrmhhuvghlsehshhholhhlrghn
 ugdrohhrgh
X-ME-Proxy: <xmx:PKXxX8aGyGAuoax6VVJBOXbpc6g9rU450vlPi4g-tYFcYcfT3pvjlA>
 <xmx:PKXxX3Wy4pXmKhSYYBz5gbFMfnnCw_rvZLhXqw9y9PIXUWT0Yhva0Q>
 <xmx:PKXxXyne-rMBTaey1BlwtWgN1HhiAExVA4J_mGP5KvuDbFTvaIHNfA>
 <xmx:PqXxX5upknzqKqXYtJpxbrXAbrDdPwDx_ccNBbXvF6idzVUxyOrkCg>
Received: from titanium.stl.sholland.net
 (70-135-148-151.lightspeed.stlsmo.sbcglobal.net [70.135.148.151])
 by mail.messagingengine.com (Postfix) with ESMTPA id 6181C24005B;
 Sun,  3 Jan 2021 06:06:36 -0500 (EST)
From: Samuel Holland <samuel@sholland.org>
To: Maxime Ripard <mripard@kernel.org>, Chen-Yu Tsai <wens@csie.org>,
 Jernej Skrabec <jernej.skrabec@siol.net>, Ondrej Jirman <megous@megous.com>
Subject: [PATCH 1/4] bus: sunxi-rsb: Move OF match table
Date: Sun,  3 Jan 2021 05:06:32 -0600
Message-Id: <20210103110635.34823-2-samuel@sholland.org>
X-Mailer: git-send-email 2.26.2
In-Reply-To: <20210103110635.34823-1-samuel@sholland.org>
References: <20210103110635.34823-1-samuel@sholland.org>
MIME-Version: 1.0
X-CRM114-Version: 20100106-BlameMichelson ( TRE 0.8.0 (BSD) ) MR-646709E3 
X-CRM114-CacheID: sfid-20210103_060645_928449_9B648C01 
X-CRM114-Status: GOOD (  10.81  )
X-BeenThere: linux-arm-kernel@lists.infradead.org
X-Mailman-Version: 2.1.29
Precedence: list
List-Id: <linux-arm-kernel.lists.infradead.org>
List-Unsubscribe: 
 <http://lists.infradead.org/mailman/options/linux-arm-kernel>,
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=unsubscribe>
List-Archive: <http://lists.infradead.org/pipermail/linux-arm-kernel/>
List-Post: <mailto:linux-arm-kernel@lists.infradead.org>
List-Help: <mailto:linux-arm-kernel-request@lists.infradead.org?subject=help>
List-Subscribe: 
 <http://lists.infradead.org/mailman/listinfo/linux-arm-kernel>,
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=subscribe>
Cc: linux-sunxi@googlegroups.com, linux-kernel@vger.kernel.org,
 linux-arm-kernel@lists.infradead.org, Samuel Holland <samuel@sholland.org>
Sender: "linux-arm-kernel" <linux-arm-kernel-bounces@lists.infradead.org>
Errors-To: 
 linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@lists.infradead.org

For some reason, this driver's OF match table was placed above the
probe/remove functions, far away from the platform_driver definition.
Adding device PM ops would move the table even farther away. Let's move
it to the usual place, right before the platform_driver.

Signed-off-by: Samuel Holland <samuel@sholland.org>
---
 drivers/bus/sunxi-rsb.c | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/drivers/bus/sunxi-rsb.c b/drivers/bus/sunxi-rsb.c
index 1bb00a959c67..c13340cab27a 100644
--- a/drivers/bus/sunxi-rsb.c
+++ b/drivers/bus/sunxi-rsb.c
@@ -614,12 +614,6 @@ static int of_rsb_register_devices(struct sunxi_rsb *rsb)
 	return 0;
 }
 
-static const struct of_device_id sunxi_rsb_of_match_table[] = {
-	{ .compatible = "allwinner,sun8i-a23-rsb" },
-	{}
-};
-MODULE_DEVICE_TABLE(of, sunxi_rsb_of_match_table);
-
 static int sunxi_rsb_probe(struct platform_device *pdev)
 {
 	struct device *dev = &pdev->dev;
@@ -747,6 +741,12 @@ static int sunxi_rsb_remove(struct platform_device *pdev)
 	return 0;
 }
 
+static const struct of_device_id sunxi_rsb_of_match_table[] = {
+	{ .compatible = "allwinner,sun8i-a23-rsb" },
+	{}
+};
+MODULE_DEVICE_TABLE(of, sunxi_rsb_of_match_table);
+
 static struct platform_driver sunxi_rsb_driver = {
 	.probe = sunxi_rsb_probe,
 	.remove	= sunxi_rsb_remove,

From patchwork Sun Jan  3 11:06:33 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Samuel Holland <samuel@sholland.org>
X-Patchwork-Id: 11995649
Return-Path: 
 <SRS0=N8P+=GG=lists.infradead.org=linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-17.0 required=3.0 tests=BAYES_00,DKIMWL_WL_HIGH,
	DKIM_SIGNED,DKIM_VALID,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,
	INCLUDES_PATCH,MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,USER_AGENT_GIT
	autolearn=unavailable autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 51641C433DB
	for <linux-arm-kernel@archiver.kernel.org>;
 Sun,  3 Jan 2021 11:08:49 +0000 (UTC)
Received: from merlin.infradead.org (merlin.infradead.org [205.233.59.134])
	(using TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by mail.kernel.org (Postfix) with ESMTPS id 143AF20771
	for <linux-arm-kernel@archiver.kernel.org>;
 Sun,  3 Jan 2021 11:08:49 +0000 (UTC)
DMARC-Filter: OpenDMARC Filter v1.3.2 mail.kernel.org 143AF20771
Authentication-Results: mail.kernel.org;
 dmarc=none (p=none dis=none) header.from=sholland.org
Authentication-Results: mail.kernel.org;
 spf=none
 smtp.mailfrom=linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@lists.infradead.org
DKIM-Signature: v=1; a=rsa-sha256; q=dns/txt; c=relaxed/relaxed;
	d=lists.infradead.org; s=merlin.20170209; h=Sender:Content-Transfer-Encoding:
	Content-Type:Cc:List-Subscribe:List-Help:List-Post:List-Archive:
	List-Unsubscribe:List-Id:MIME-Version:References:In-Reply-To:Message-Id:Date:
	Subject:To:From:Reply-To:Content-ID:Content-Description:Resent-Date:
	Resent-From:Resent-Sender:Resent-To:Resent-Cc:Resent-Message-ID:List-Owner;
	 bh=HDSwIvUBM1f9wb9RJolisi7g/oopy1dQ3XUMpiKvvPM=; b=xJeYdL39YXB9CNX51d3OFCu6h
	Z1w5erLVotjfBRlCitJZwv0Z/LsNNXITIFE/ttpEGn45tDGBI1+DPv7v9qotNIOeP8T47k+owkHNy
	9R4m8KDyxHbUdRMDNEEujE2NfjbaD8fkl8v2Uv+vV3SuOF4kReIUyjTO1e0vc8Ug4Tqy4VTxPwgEY
	Xi0xo4rJdpap8Xi36YmY6E/atxehL3fo6mxMffR6ztUVNW+2L1FLZY6CqRwr2cx2nrX1X+OCTaK5+
	QyeCxGmxZAytuGdYdf5DFnItxjsDwjA2lscILlyM1XWK86MtadLOnHsNnt0QWMMgkKcb8j+75Mmuv
	vdINjIGvQ==;
Received: from localhost ([::1] helo=merlin.infradead.org)
	by merlin.infradead.org with esmtp (Exim 4.92.3 #3 (Red Hat Linux))
	id 1kw1Dr-0002S1-Hh; Sun, 03 Jan 2021 11:06:55 +0000
Received: from wout5-smtp.messagingengine.com ([64.147.123.21])
 by merlin.infradead.org with esmtps (Exim 4.92.3 #3 (Red Hat Linux))
 id 1kw1Dh-0002Nm-Jc
 for linux-arm-kernel@lists.infradead.org; Sun, 03 Jan 2021 11:06:48 +0000
Received: from compute5.internal (compute5.nyi.internal [10.202.2.45])
 by mailout.west.internal (Postfix) with ESMTP id AF126514;
 Sun,  3 Jan 2021 06:06:38 -0500 (EST)
Received: from mailfrontend1 ([10.202.2.162])
 by compute5.internal (MEProxy); Sun, 03 Jan 2021 06:06:39 -0500
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=sholland.org; h=
 from:to:cc:subject:date:message-id:in-reply-to:references
 :mime-version:content-transfer-encoding; s=fm1; bh=7SvhTIDFXD6wc
 1Hcq2TPDMxEHxB+DmYYUXL05Wstfc8=; b=SSM6G6BRvnEacDEyLPtxk4J1fLMku
 8k5GeCL7ttGGcKBCETs6KFhT9mV6RXumJfaWZzL9Cwz+EQlQJZsdZpGKmUI2xL7A
 9QZoiRgDiVXJZRG8Ml4zVBrQHmn/5vibAmqmRkF5WMgj0k/19z0RfvC8zAJlAgAG
 LFNfu3mpVOxZxdgTEqIuy8otmw5L8Kgvf4Pmf27JH9W/WL/TnlTObmNBOm2DLygC
 MAJsNvKvlZU/hcSXfagkarH3s66wi8/HTOg51qLLxYSsbPY7A89DmvKoTNSdohMh
 JbgNSMsjPg8JT1sYRhebE0XAAZeiO2htIo8aoYxC6stUp7zhIvWEeuN/w==
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=
 messagingengine.com; h=cc:content-transfer-encoding:date:from
 :in-reply-to:message-id:mime-version:references:subject:to
 :x-me-proxy:x-me-proxy:x-me-sender:x-me-sender:x-sasl-enc; s=
 fm1; bh=7SvhTIDFXD6wc1Hcq2TPDMxEHxB+DmYYUXL05Wstfc8=; b=acrkg0Fm
 iI9N1ScvTG9oC0/1K30sKXAPO6qqq0t4Pmftd2yLA4itcEWPgriW6JHMVpHn5KCi
 YNFUNSsB9lmBtR6KKBTpzktMQOZSs56kLUZtWb8klRAWi7Apomb/Y89W5JaWx4Zb
 xWapKgUZto9wAi7F7/Hcuqi8jaI6jzf8KGIwc+l98MJG6isTJDHePYDh1bAGAPc5
 kLXzZ32PEK0NIw/yk+yTEcWl1eTCJ+4AQpRnUzp0WZsLW48IG/DeM57GOCOyjwdS
 RhrJYUla9zIq9T73nAkoNLwwR7y9YbBl7LWm5hmDg/SmWn+NWcxiFW2EXoxjIJKl
 yuc94ekw4LiqEQ==
X-ME-Sender: <xms:PaXxX1T0vc_wOGCBviEpK45rKaFwFK6koZtrtvNbiC-rKoh1m51pIA>
 <xme:PaXxX-ybynqMABVPiYLZateHQJ_LqzeCxmlZrdJsEr9ezS2TLFLLynYsvSIqAPgdP
 foOp6Kl7hLpG0lmng>
X-ME-Proxy-Cause: 
 gggruggvucftvghtrhhoucdtuddrgedujedrvdefuddgvdegucetufdoteggodetrfdotf
 fvucfrrhhofhhilhgvmecuhfgrshhtofgrihhlpdfqfgfvpdfurfetoffkrfgpnffqhgen
 uceurghilhhouhhtmecufedttdenucesvcftvggtihhpihgvnhhtshculddquddttddmne
 cujfgurhephffvufffkffojghfggfgsedtkeertdertddtnecuhfhrohhmpefurghmuhgv
 lhcujfholhhlrghnugcuoehsrghmuhgvlhesshhhohhllhgrnhgurdhorhhgqeenucggtf
 frrghtthgvrhhnpeduhfejfedvhffgfeehtefghfeiiefgfeehgfdvvdevfeegjeehjedv
 gfejheeuieenucfkphepjedtrddufeehrddugeekrdduhedunecuvehluhhsthgvrhfuih
 iivgeptdenucfrrghrrghmpehmrghilhhfrhhomhepshgrmhhuvghlsehshhholhhlrghn
 ugdrohhrgh
X-ME-Proxy: <xmx:PaXxX63DSq2lFrKREm3q-j_SVWCpu-Lfm8sB59cOnlezoT8qMMROhw>
 <xmx:PaXxX9AqAqPiUrPcH927xs0wgDb0qgPffsvVhhBLWvfKjNQQV9sjFw>
 <xmx:PaXxX-jeZm7_7BPvFhuN7k5Z0lLA_51mNnOleHqN0-Y6jOFjc4N4kA>
 <xmx:PqXxXwajSs33lbK3ploMAQkcXaIp98gqFIcJpBSzBfetVZ1yPuYtFw>
Received: from titanium.stl.sholland.net
 (70-135-148-151.lightspeed.stlsmo.sbcglobal.net [70.135.148.151])
 by mail.messagingengine.com (Postfix) with ESMTPA id BC83F24005D;
 Sun,  3 Jan 2021 06:06:36 -0500 (EST)
From: Samuel Holland <samuel@sholland.org>
To: Maxime Ripard <mripard@kernel.org>, Chen-Yu Tsai <wens@csie.org>,
 Jernej Skrabec <jernej.skrabec@siol.net>, Ondrej Jirman <megous@megous.com>
Subject: [PATCH 2/4] bus: sunxi-rsb: Split out controller init/exit functions
Date: Sun,  3 Jan 2021 05:06:33 -0600
Message-Id: <20210103110635.34823-3-samuel@sholland.org>
X-Mailer: git-send-email 2.26.2
In-Reply-To: <20210103110635.34823-1-samuel@sholland.org>
References: <20210103110635.34823-1-samuel@sholland.org>
MIME-Version: 1.0
X-CRM114-Version: 20100106-BlameMichelson ( TRE 0.8.0 (BSD) ) MR-646709E3 
X-CRM114-CacheID: sfid-20210103_060645_956895_0E1C29FC 
X-CRM114-Status: GOOD (  18.07  )
X-BeenThere: linux-arm-kernel@lists.infradead.org
X-Mailman-Version: 2.1.29
Precedence: list
List-Id: <linux-arm-kernel.lists.infradead.org>
List-Unsubscribe: 
 <http://lists.infradead.org/mailman/options/linux-arm-kernel>,
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=unsubscribe>
List-Archive: <http://lists.infradead.org/pipermail/linux-arm-kernel/>
List-Post: <mailto:linux-arm-kernel@lists.infradead.org>
List-Help: <mailto:linux-arm-kernel-request@lists.infradead.org?subject=help>
List-Subscribe: 
 <http://lists.infradead.org/mailman/listinfo/linux-arm-kernel>,
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=subscribe>
Cc: linux-sunxi@googlegroups.com, linux-kernel@vger.kernel.org,
 linux-arm-kernel@lists.infradead.org, Samuel Holland <samuel@sholland.org>
Sender: "linux-arm-kernel" <linux-arm-kernel-bounces@lists.infradead.org>
Errors-To: 
 linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@lists.infradead.org

This separates the resource acquisition from the hardware initialization
phase, so the hardware initialization can be repeated after system
suspend/resume. The same is done for the exit/remove function, except
that there is no resource deallocation phase due to the use of devres.

The requested RSB clock frequency is stored in `struct sunxi_rsb` so it
will be available when reinitializing the hardware.

Signed-off-by: Samuel Holland <samuel@sholland.org>
---
 drivers/bus/sunxi-rsb.c | 127 ++++++++++++++++++++++------------------
 1 file changed, 71 insertions(+), 56 deletions(-)

diff --git a/drivers/bus/sunxi-rsb.c b/drivers/bus/sunxi-rsb.c
index c13340cab27a..3f290da45619 100644
--- a/drivers/bus/sunxi-rsb.c
+++ b/drivers/bus/sunxi-rsb.c
@@ -126,6 +126,7 @@ struct sunxi_rsb {
 	struct completion complete;
 	struct mutex lock;
 	unsigned int status;
+	u32 clk_freq;
 };
 
 /* bus / slave device related functions */
@@ -614,16 +615,74 @@ static int of_rsb_register_devices(struct sunxi_rsb *rsb)
 	return 0;
 }
 
+static int sunxi_rsb_hw_init(struct sunxi_rsb *rsb)
+{
+	struct device *dev = rsb->dev;
+	unsigned long p_clk_freq;
+	u32 clk_delay, reg;
+	int clk_div, ret;
+
+	ret = clk_prepare_enable(rsb->clk);
+	if (ret) {
+		dev_err(dev, "failed to enable clk: %d\n", ret);
+		return ret;
+	}
+
+	ret = reset_control_deassert(rsb->rstc);
+	if (ret) {
+		dev_err(dev, "failed to deassert reset line: %d\n", ret);
+		goto err_clk_disable;
+	}
+
+	/* reset the controller */
+	writel(RSB_CTRL_SOFT_RST, rsb->regs + RSB_CTRL);
+	readl_poll_timeout(rsb->regs + RSB_CTRL, reg,
+			   !(reg & RSB_CTRL_SOFT_RST), 1000, 100000);
+
+	/*
+	 * Clock frequency and delay calculation code is from
+	 * Allwinner U-boot sources.
+	 *
+	 * From A83 user manual:
+	 * bus clock frequency = parent clock frequency / (2 * (divider + 1))
+	 */
+	p_clk_freq = clk_get_rate(rsb->clk);
+	clk_div = p_clk_freq / rsb->clk_freq / 2;
+	if (!clk_div)
+		clk_div = 1;
+	else if (clk_div > RSB_CCR_MAX_CLK_DIV + 1)
+		clk_div = RSB_CCR_MAX_CLK_DIV + 1;
+
+	clk_delay = clk_div >> 1;
+	if (!clk_delay)
+		clk_delay = 1;
+
+	dev_info(dev, "RSB running at %lu Hz\n", p_clk_freq / clk_div / 2);
+	writel(RSB_CCR_SDA_OUT_DELAY(clk_delay) | RSB_CCR_CLK_DIV(clk_div - 1),
+	       rsb->regs + RSB_CCR);
+
+	return 0;
+
+err_clk_disable:
+	clk_disable_unprepare(rsb->clk);
+
+	return ret;
+}
+
+static void sunxi_rsb_hw_exit(struct sunxi_rsb *rsb)
+{
+	reset_control_assert(rsb->rstc);
+	clk_disable_unprepare(rsb->clk);
+}
+
 static int sunxi_rsb_probe(struct platform_device *pdev)
 {
 	struct device *dev = &pdev->dev;
 	struct device_node *np = dev->of_node;
 	struct resource *r;
 	struct sunxi_rsb *rsb;
-	unsigned long p_clk_freq;
-	u32 clk_delay, clk_freq = 3000000;
-	int clk_div, irq, ret;
-	u32 reg;
+	u32 clk_freq = 3000000;
+	int irq, ret;
 
 	of_property_read_u32(np, "clock-frequency", &clk_freq);
 	if (clk_freq > RSB_MAX_FREQ) {
@@ -638,6 +697,7 @@ static int sunxi_rsb_probe(struct platform_device *pdev)
 		return -ENOMEM;
 
 	rsb->dev = dev;
+	rsb->clk_freq = clk_freq;
 	platform_set_drvdata(pdev, rsb);
 	r = platform_get_resource(pdev, IORESOURCE_MEM, 0);
 	rsb->regs = devm_ioremap_resource(dev, r);
@@ -655,63 +715,27 @@ static int sunxi_rsb_probe(struct platform_device *pdev)
 		return ret;
 	}
 
-	ret = clk_prepare_enable(rsb->clk);
-	if (ret) {
-		dev_err(dev, "failed to enable clk: %d\n", ret);
-		return ret;
-	}
-
-	p_clk_freq = clk_get_rate(rsb->clk);
-
 	rsb->rstc = devm_reset_control_get(dev, NULL);
 	if (IS_ERR(rsb->rstc)) {
 		ret = PTR_ERR(rsb->rstc);
 		dev_err(dev, "failed to retrieve reset controller: %d\n", ret);
-		goto err_clk_disable;
-	}
-
-	ret = reset_control_deassert(rsb->rstc);
-	if (ret) {
-		dev_err(dev, "failed to deassert reset line: %d\n", ret);
-		goto err_clk_disable;
+		return ret;
 	}
 
 	init_completion(&rsb->complete);
 	mutex_init(&rsb->lock);
 
-	/* reset the controller */
-	writel(RSB_CTRL_SOFT_RST, rsb->regs + RSB_CTRL);
-	readl_poll_timeout(rsb->regs + RSB_CTRL, reg,
-			   !(reg & RSB_CTRL_SOFT_RST), 1000, 100000);
-
-	/*
-	 * Clock frequency and delay calculation code is from
-	 * Allwinner U-boot sources.
-	 *
-	 * From A83 user manual:
-	 * bus clock frequency = parent clock frequency / (2 * (divider + 1))
-	 */
-	clk_div = p_clk_freq / clk_freq / 2;
-	if (!clk_div)
-		clk_div = 1;
-	else if (clk_div > RSB_CCR_MAX_CLK_DIV + 1)
-		clk_div = RSB_CCR_MAX_CLK_DIV + 1;
-
-	clk_delay = clk_div >> 1;
-	if (!clk_delay)
-		clk_delay = 1;
-
-	dev_info(dev, "RSB running at %lu Hz\n", p_clk_freq / clk_div / 2);
-	writel(RSB_CCR_SDA_OUT_DELAY(clk_delay) | RSB_CCR_CLK_DIV(clk_div - 1),
-	       rsb->regs + RSB_CCR);
-
 	ret = devm_request_irq(dev, irq, sunxi_rsb_irq, 0, RSB_CTRL_NAME, rsb);
 	if (ret) {
 		dev_err(dev, "can't register interrupt handler irq %d: %d\n",
 			irq, ret);
-		goto err_reset_assert;
+		return ret;
 	}
 
+	ret = sunxi_rsb_hw_init(rsb);
+	if (ret)
+		return ret;
+
 	/* initialize all devices on the bus into RSB mode */
 	ret = sunxi_rsb_init_device_mode(rsb);
 	if (ret)
@@ -720,14 +744,6 @@ static int sunxi_rsb_probe(struct platform_device *pdev)
 	of_rsb_register_devices(rsb);
 
 	return 0;
-
-err_reset_assert:
-	reset_control_assert(rsb->rstc);
-
-err_clk_disable:
-	clk_disable_unprepare(rsb->clk);
-
-	return ret;
 }
 
 static int sunxi_rsb_remove(struct platform_device *pdev)
@@ -735,8 +751,7 @@ static int sunxi_rsb_remove(struct platform_device *pdev)
 	struct sunxi_rsb *rsb = platform_get_drvdata(pdev);
 
 	device_for_each_child(rsb->dev, NULL, sunxi_rsb_remove_devices);
-	reset_control_assert(rsb->rstc);
-	clk_disable_unprepare(rsb->clk);
+	sunxi_rsb_hw_exit(rsb);
 
 	return 0;
 }

From patchwork Sun Jan  3 11:06:34 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Samuel Holland <samuel@sholland.org>
X-Patchwork-Id: 11995643
Return-Path: 
 <SRS0=N8P+=GG=lists.infradead.org=linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-17.0 required=3.0 tests=BAYES_00,DKIMWL_WL_HIGH,
	DKIM_SIGNED,DKIM_VALID,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,
	INCLUDES_PATCH,MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,USER_AGENT_GIT
	autolearn=unavailable autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id B8C5FC433E6
	for <linux-arm-kernel@archiver.kernel.org>;
 Sun,  3 Jan 2021 11:08:45 +0000 (UTC)
Received: from merlin.infradead.org (merlin.infradead.org [205.233.59.134])
	(using TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by mail.kernel.org (Postfix) with ESMTPS id 7B3C5207FB
	for <linux-arm-kernel@archiver.kernel.org>;
 Sun,  3 Jan 2021 11:08:45 +0000 (UTC)
DMARC-Filter: OpenDMARC Filter v1.3.2 mail.kernel.org 7B3C5207FB
Authentication-Results: mail.kernel.org;
 dmarc=none (p=none dis=none) header.from=sholland.org
Authentication-Results: mail.kernel.org;
 spf=none
 smtp.mailfrom=linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@lists.infradead.org
DKIM-Signature: v=1; a=rsa-sha256; q=dns/txt; c=relaxed/relaxed;
	d=lists.infradead.org; s=merlin.20170209; h=Sender:Content-Transfer-Encoding:
	Content-Type:Cc:List-Subscribe:List-Help:List-Post:List-Archive:
	List-Unsubscribe:List-Id:MIME-Version:References:In-Reply-To:Message-Id:Date:
	Subject:To:From:Reply-To:Content-ID:Content-Description:Resent-Date:
	Resent-From:Resent-Sender:Resent-To:Resent-Cc:Resent-Message-ID:List-Owner;
	 bh=aX+r7EtwAd6FxZZAoHmaYNLBPIAQVMY6nY9Qei6NJto=; b=fKTierKhcS9t+MiyzWULkWNYD
	guaKYoOumiJCtQJR5r1GhyPavXctxYFoJqFWhOJfoCvmBpPO4TCGTf2AIMBAyG1ecly4wfth3fOV/
	O2aqrGhD62nMY2j+zZ7xFnkUQsIqUQz7ymnwOiFCcogCkJTNzj0FTAKnlQSC/6kLn3ZOh6MxS4xjR
	JwqwdIUmMJ7wO/DL0VJ5sRgZTG7PbdyZejWcyw+tTY0lArhxmKjwgnE7sqLjI7BEoXYdl3mWie9xZ
	25z9/F13wurSWPjz396PjeLBWTAeL9gcK+CjYX1v8aJWjMVp+THt5oKSywAYwPgPdgMo8jnDeY3Vx
	9WZy0XYqQ==;
Received: from localhost ([::1] helo=merlin.infradead.org)
	by merlin.infradead.org with esmtp (Exim 4.92.3 #3 (Red Hat Linux))
	id 1kw1Dm-0002Qk-Gh; Sun, 03 Jan 2021 11:06:50 +0000
Received: from wout5-smtp.messagingengine.com ([64.147.123.21])
 by merlin.infradead.org with esmtps (Exim 4.92.3 #3 (Red Hat Linux))
 id 1kw1Dh-0002Np-Jb
 for linux-arm-kernel@lists.infradead.org; Sun, 03 Jan 2021 11:06:46 +0000
Received: from compute5.internal (compute5.nyi.internal [10.202.2.45])
 by mailout.west.internal (Postfix) with ESMTP id C5760542;
 Sun,  3 Jan 2021 06:06:38 -0500 (EST)
Received: from mailfrontend1 ([10.202.2.162])
 by compute5.internal (MEProxy); Sun, 03 Jan 2021 06:06:39 -0500
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=sholland.org; h=
 from:to:cc:subject:date:message-id:in-reply-to:references
 :mime-version:content-transfer-encoding; s=fm1; bh=eiXmcA+mRsHUH
 aYx5zquPjpU06fK14Zb7hMHBXjnmy8=; b=amYS3bs3zSCXVyHRUQm64I2JXGUjy
 kzs4a6MoHvoyY0ylnkvzOAgUATaYmmNZLT6cpsIWn95Hs8/XFOQTSblwQdF5mEwc
 i3bS4bUrxIO3Ss9vkuq/jUCe58xHPIipjeIzMVf18WHYOogEw1hRXlGbIZRLcP7t
 +wQa7oBby8Dkia+iFwGKoVlHz3g1zuP4X8wtUqHKigf7bd7oUG3HrneeFUKPZHpo
 ciHnAvtU4+QNTMe+U1hl+/nJ65dGmhPjCwSTx0YrdXFJtCIMvp7thhpXcef2gtw6
 HOdmO5WJD+UTJ2PhJaG8qZaFOs9XI+SNoafLd4t6w/1UEw0EGrEmgpODg==
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=
 messagingengine.com; h=cc:content-transfer-encoding:date:from
 :in-reply-to:message-id:mime-version:references:subject:to
 :x-me-proxy:x-me-proxy:x-me-sender:x-me-sender:x-sasl-enc; s=
 fm1; bh=eiXmcA+mRsHUHaYx5zquPjpU06fK14Zb7hMHBXjnmy8=; b=AT+16bHI
 qzNOfY8w/++x4R0+Zt3H7qDDracIhDEDbl+r2ESVlH5GJM+mqyztW0mXok8udn/R
 yrcfULwrLpzOTtRr06v4xNvcSoUc/p/+RAv+lrxiv6eVZwSf4QUxKRE4fLHWXSRg
 uP0KYeWvKgwWKXsgdnUzELkbl03bukGNRniGHB/SpR3sNfrZmYAPVn5vh2f9FmWT
 U93FQ8HXztIKUkqGKuLseMFWouvnjLFVqjQ0F27yF6TEWtRv+v6BrDcE0D2hD9OS
 V3UvfUZsOAt5qyKXB6meP111YXLll2xWFIfCCF7oALjmAE+vDZtPUYkkommCaubL
 7QwIf7Qz/zxCWA==
X-ME-Sender: <xms:PaXxX3WiaoNB27KVVSAGN03DxKdFspTzX41EtACxGzK5xCSiW-xuSw>
 <xme:PaXxX_llLkSEWUl9cWGB4yTpMrrozfy7fx3fe0chJARjdBDH6gYmR9_cg_hmxfMq8
 BvEvalVxQMWip26Mw>
X-ME-Proxy-Cause: 
 gggruggvucftvghtrhhoucdtuddrgedujedrvdefuddgvdegucetufdoteggodetrfdotf
 fvucfrrhhofhhilhgvmecuhfgrshhtofgrihhlpdfqfgfvpdfurfetoffkrfgpnffqhgen
 uceurghilhhouhhtmecufedttdenucesvcftvggtihhpihgvnhhtshculddquddttddmne
 cujfgurhephffvufffkffojghfggfgsedtkeertdertddtnecuhfhrohhmpefurghmuhgv
 lhcujfholhhlrghnugcuoehsrghmuhgvlhesshhhohhllhgrnhgurdhorhhgqeenucggtf
 frrghtthgvrhhnpeduhfejfedvhffgfeehtefghfeiiefgfeehgfdvvdevfeegjeehjedv
 gfejheeuieenucfkphepjedtrddufeehrddugeekrdduhedunecuvehluhhsthgvrhfuih
 iivgeptdenucfrrghrrghmpehmrghilhhfrhhomhepshgrmhhuvghlsehshhholhhlrghn
 ugdrohhrgh
X-ME-Proxy: <xmx:PaXxXzZ5S55n9nTlxEKG5njACIXfS5pDpGweAfrRHYz7j1nOOluIvg>
 <xmx:PaXxXyWMtAvDftKQPZqijj-1rty5W1Kbj9fJtXDSAiK6i6Z4Z4-Dfw>
 <xmx:PaXxXxmuKLZI4hALCFhsp3O8Ad7Quj7PFy2Vg1bM5BRdstI4KBOvPA>
 <xmx:PqXxX0tUHKlFrwXUyL5Sl7TEIKT62yLLqGA5h50SN382DEoNr0wl_Q>
Received: from titanium.stl.sholland.net
 (70-135-148-151.lightspeed.stlsmo.sbcglobal.net [70.135.148.151])
 by mail.messagingengine.com (Postfix) with ESMTPA id 22FF524005E;
 Sun,  3 Jan 2021 06:06:37 -0500 (EST)
From: Samuel Holland <samuel@sholland.org>
To: Maxime Ripard <mripard@kernel.org>, Chen-Yu Tsai <wens@csie.org>,
 Jernej Skrabec <jernej.skrabec@siol.net>, Ondrej Jirman <megous@megous.com>
Subject: [PATCH 3/4] bus: sunxi-rsb: Implement suspend/resume/shutdown
 callbacks
Date: Sun,  3 Jan 2021 05:06:34 -0600
Message-Id: <20210103110635.34823-4-samuel@sholland.org>
X-Mailer: git-send-email 2.26.2
In-Reply-To: <20210103110635.34823-1-samuel@sholland.org>
References: <20210103110635.34823-1-samuel@sholland.org>
MIME-Version: 1.0
X-CRM114-Version: 20100106-BlameMichelson ( TRE 0.8.0 (BSD) ) MR-646709E3 
X-CRM114-CacheID: sfid-20210103_060645_882346_BEC20F57 
X-CRM114-Status: GOOD (  14.12  )
X-BeenThere: linux-arm-kernel@lists.infradead.org
X-Mailman-Version: 2.1.29
Precedence: list
List-Id: <linux-arm-kernel.lists.infradead.org>
List-Unsubscribe: 
 <http://lists.infradead.org/mailman/options/linux-arm-kernel>,
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=unsubscribe>
List-Archive: <http://lists.infradead.org/pipermail/linux-arm-kernel/>
List-Post: <mailto:linux-arm-kernel@lists.infradead.org>
List-Help: <mailto:linux-arm-kernel-request@lists.infradead.org?subject=help>
List-Subscribe: 
 <http://lists.infradead.org/mailman/listinfo/linux-arm-kernel>,
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=subscribe>
Cc: linux-sunxi@googlegroups.com, linux-kernel@vger.kernel.org,
 linux-arm-kernel@lists.infradead.org, Samuel Holland <samuel@sholland.org>
Sender: "linux-arm-kernel" <linux-arm-kernel-bounces@lists.infradead.org>
Errors-To: 
 linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@lists.infradead.org

Since system firmware is likely to use the RSB bus to communicate with a
PMIC while the system is suspended, we cannot make any assumptions about
the controller state after resuming. Thus it is important to completely
reinitialize the controller.

The RSB bus needs to be ready as soon as IRQs are enabled, to handle
wakeup event IRQs coming from the PMIC. Thus it uses NOIRQ callbacks.

Signed-off-by: Samuel Holland <samuel@sholland.org>
---
 drivers/bus/sunxi-rsb.c | 30 ++++++++++++++++++++++++++++++
 1 file changed, 30 insertions(+)

diff --git a/drivers/bus/sunxi-rsb.c b/drivers/bus/sunxi-rsb.c
index 3f290da45619..efd222f36cdc 100644
--- a/drivers/bus/sunxi-rsb.c
+++ b/drivers/bus/sunxi-rsb.c
@@ -45,6 +45,7 @@
 #include <linux/of_irq.h>
 #include <linux/of_platform.h>
 #include <linux/platform_device.h>
+#include <linux/pm.h>
 #include <linux/regmap.h>
 #include <linux/reset.h>
 #include <linux/slab.h>
@@ -675,6 +676,22 @@ static void sunxi_rsb_hw_exit(struct sunxi_rsb *rsb)
 	clk_disable_unprepare(rsb->clk);
 }
 
+static int __maybe_unused sunxi_rsb_suspend(struct device *dev)
+{
+	struct sunxi_rsb *rsb = dev_get_drvdata(dev);
+
+	sunxi_rsb_hw_exit(rsb);
+
+	return 0;
+}
+
+static int __maybe_unused sunxi_rsb_resume(struct device *dev)
+{
+	struct sunxi_rsb *rsb = dev_get_drvdata(dev);
+
+	return sunxi_rsb_hw_init(rsb);
+}
+
 static int sunxi_rsb_probe(struct platform_device *pdev)
 {
 	struct device *dev = &pdev->dev;
@@ -756,6 +773,17 @@ static int sunxi_rsb_remove(struct platform_device *pdev)
 	return 0;
 }
 
+static void sunxi_rsb_shutdown(struct platform_device *pdev)
+{
+	struct sunxi_rsb *rsb = platform_get_drvdata(pdev);
+
+	sunxi_rsb_hw_exit(rsb);
+}
+
+static const struct dev_pm_ops sunxi_rsb_dev_pm_ops = {
+	SET_NOIRQ_SYSTEM_SLEEP_PM_OPS(sunxi_rsb_suspend, sunxi_rsb_resume)
+};
+
 static const struct of_device_id sunxi_rsb_of_match_table[] = {
 	{ .compatible = "allwinner,sun8i-a23-rsb" },
 	{}
@@ -765,9 +793,11 @@ MODULE_DEVICE_TABLE(of, sunxi_rsb_of_match_table);
 static struct platform_driver sunxi_rsb_driver = {
 	.probe = sunxi_rsb_probe,
 	.remove	= sunxi_rsb_remove,
+	.shutdown = sunxi_rsb_shutdown,
 	.driver	= {
 		.name = RSB_CTRL_NAME,
 		.of_match_table = sunxi_rsb_of_match_table,
+		.pm = &sunxi_rsb_dev_pm_ops,
 	},
 };
 

From patchwork Sun Jan  3 11:06:35 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Samuel Holland <samuel@sholland.org>
X-Patchwork-Id: 11995645
Return-Path: 
 <SRS0=N8P+=GG=lists.infradead.org=linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-14.2 required=3.0 tests=BAYES_00,DKIMWL_WL_HIGH,
	DKIM_SIGNED,DKIM_VALID,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,
	INCLUDES_PATCH,MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,
	UNWANTED_LANGUAGE_BODY,USER_AGENT_GIT autolearn=unavailable
	autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id B44E5C433DB
	for <linux-arm-kernel@archiver.kernel.org>;
 Sun,  3 Jan 2021 11:08:45 +0000 (UTC)
Received: from merlin.infradead.org (merlin.infradead.org [205.233.59.134])
	(using TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by mail.kernel.org (Postfix) with ESMTPS id 71EF920771
	for <linux-arm-kernel@archiver.kernel.org>;
 Sun,  3 Jan 2021 11:08:45 +0000 (UTC)
DMARC-Filter: OpenDMARC Filter v1.3.2 mail.kernel.org 71EF920771
Authentication-Results: mail.kernel.org;
 dmarc=none (p=none dis=none) header.from=sholland.org
Authentication-Results: mail.kernel.org;
 spf=none
 smtp.mailfrom=linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@lists.infradead.org
DKIM-Signature: v=1; a=rsa-sha256; q=dns/txt; c=relaxed/relaxed;
	d=lists.infradead.org; s=merlin.20170209; h=Sender:Content-Transfer-Encoding:
	Content-Type:Cc:List-Subscribe:List-Help:List-Post:List-Archive:
	List-Unsubscribe:List-Id:MIME-Version:References:In-Reply-To:Message-Id:Date:
	Subject:To:From:Reply-To:Content-ID:Content-Description:Resent-Date:
	Resent-From:Resent-Sender:Resent-To:Resent-Cc:Resent-Message-ID:List-Owner;
	 bh=PH3Lr/62tp12K7Tz2ZhPhbNvYhOs97sm61YHrJoTij8=; b=OmyMGVYDMm37pIPm0QleMZFSO
	s2G3pEv76uuSxwXX7xSRs+T7zSQjQ+SZSLbnxwXSlW+R9hC9CQvrfMgOLUOjRTSJSC0P7liNrUqEz
	Uz/WGF/9Bm2ps+UECklcDvo7pWGvCtqQNDamKCHxnm1GFzHGdMBZN6RtG8CuuPEPqpo/Sa6D2fhU+
	os3ILkrr19PAoLLgJqiJ6UGa/NAoVgDR//KLEhyGgVP/GPl3qpoWEKKmL7meV2vHI64LoYXpMFoef
	QZWNOaixznEmbD9Mv+2a6PTM7seyDSCKF5+cwZmog1b/ZW1fhLxpn4BhQ8KKMT7J76yUOX7EmxuOS
	JY633/uKg==;
Received: from localhost ([::1] helo=merlin.infradead.org)
	by merlin.infradead.org with esmtp (Exim 4.92.3 #3 (Red Hat Linux))
	id 1kw1Dq-0002Rg-9w; Sun, 03 Jan 2021 11:06:54 +0000
Received: from wout5-smtp.messagingengine.com ([64.147.123.21])
 by merlin.infradead.org with esmtps (Exim 4.92.3 #3 (Red Hat Linux))
 id 1kw1Dh-0002Nq-Jb
 for linux-arm-kernel@lists.infradead.org; Sun, 03 Jan 2021 11:06:48 +0000
Received: from compute5.internal (compute5.nyi.internal [10.202.2.45])
 by mailout.west.internal (Postfix) with ESMTP id A11D6275;
 Sun,  3 Jan 2021 06:06:38 -0500 (EST)
Received: from mailfrontend1 ([10.202.2.162])
 by compute5.internal (MEProxy); Sun, 03 Jan 2021 06:06:39 -0500
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=sholland.org; h=
 from:to:cc:subject:date:message-id:in-reply-to:references
 :mime-version:content-transfer-encoding; s=fm1; bh=jb/i2UCQx+8gx
 zanVJoTiDULrP/D/wg+dgI19BaOfYw=; b=PpIotdJdBm+IVR5zlxpp2B3n9CJl5
 fuRVd5cjfAsauduynj0neXoA5X7bmNOpu51woYTKrPSR8sBxKOlAe73AuRnEEwey
 M4SpBPujimDLyRxkjHp3g2gMgwTDK59OxEbkoDjUTHvT016CY+28BT9eB37SdeLn
 qrV12aMfMPjBHOTkWBaNc3sLR6yKcLauy6WOHtK9zkHwb6fLJErkV+e+BhW17EgG
 cWFjIRMVozUPEaH13sBPPavd6rsYJOBSLc8s9cHR8CVItjJ7q1vGFNLigP6INfXh
 RnZiLOvliC1YdVFQzQFGgiNcEwg9F2Ai9e3JfZLbsuS32F2ZkVNX3FqXw==
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=
 messagingengine.com; h=cc:content-transfer-encoding:date:from
 :in-reply-to:message-id:mime-version:references:subject:to
 :x-me-proxy:x-me-proxy:x-me-sender:x-me-sender:x-sasl-enc; s=
 fm1; bh=jb/i2UCQx+8gxzanVJoTiDULrP/D/wg+dgI19BaOfYw=; b=UeKRgcDk
 SIJTG1gS4o7ajgL92w0LF+Ow7Zv42PKUjNK+2C53YqZH+BJljDmo8gC651zWlHsC
 lRUF/ZJoHCtkm9kRxjLktzd2Wnpa/8puQkh21ZXaKwDSFSma3Y0Rosg79jn1Ylzb
 Csar6zZtbXLaxhIpRR5W9MtSNfgK1qlHQwRXYkAMM6d9ycrL70ms7CN0itR/c8kq
 iu72esYwND960nBQIjIpvvOOTgjZnLjEB07djsT9xT2/+/7VehtGNmNDEQ3jZIlz
 wqlw0tW98cMJfF+aBkao5stJudDJ+VZ3Z3IHrldItPE8V0zXPwCgMGVFTMnIhSyK
 t9sKkh624J0mKA==
X-ME-Sender: <xms:PaXxXxjg-KFxymogylX9XBeWdpztJQnaYfFzJhBY1jMypLQu7SLk_w>
 <xme:PaXxX2AWYz5bwLeqT6PUQpF_DQlnJAs01yPjdbeWkO1UQqlTheNcGKUdPlWrZg-tG
 3jzhg6zAkIMr44plA>
X-ME-Proxy-Cause: 
 gggruggvucftvghtrhhoucdtuddrgedujedrvdefuddgvdegucetufdoteggodetrfdotf
 fvucfrrhhofhhilhgvmecuhfgrshhtofgrihhlpdfqfgfvpdfurfetoffkrfgpnffqhgen
 uceurghilhhouhhtmecufedttdenucesvcftvggtihhpihgvnhhtshculddquddttddmne
 cujfgurhephffvufffkffojghfggfgsedtkeertdertddtnecuhfhrohhmpefurghmuhgv
 lhcujfholhhlrghnugcuoehsrghmuhgvlhesshhhohhllhgrnhgurdhorhhgqeenucggtf
 frrghtthgvrhhnpeduhfejfedvhffgfeehtefghfeiiefgfeehgfdvvdevfeegjeehjedv
 gfejheeuieenucfkphepjedtrddufeehrddugeekrdduhedunecuvehluhhsthgvrhfuih
 iivgeptdenucfrrghrrghmpehmrghilhhfrhhomhepshgrmhhuvghlsehshhholhhlrghn
 ugdrohhrgh
X-ME-Proxy: <xmx:PaXxXxEfK24cnoPaAFTBHlEBp93GObxptLXpXT2aaa6N4PlJJHnz3Q>
 <xmx:PaXxX2Ts1EzfnrugD_8DGpOSnMxJPeVXPfxg0ZobB9fBL-EXBr66Cg>
 <xmx:PaXxX-xCfuPjmFYFVOu_BFYX9cNAJsN98o-at8ZAem4NsIj008FieA>
 <xmx:PqXxX1rwSUQxLVn7MfzeNhrtB5hXLaV7UMmTfpOKsp8e2s8z1dmhdg>
Received: from titanium.stl.sholland.net
 (70-135-148-151.lightspeed.stlsmo.sbcglobal.net [70.135.148.151])
 by mail.messagingengine.com (Postfix) with ESMTPA id 7E125240062;
 Sun,  3 Jan 2021 06:06:37 -0500 (EST)
From: Samuel Holland <samuel@sholland.org>
To: Maxime Ripard <mripard@kernel.org>, Chen-Yu Tsai <wens@csie.org>,
 Jernej Skrabec <jernej.skrabec@siol.net>, Ondrej Jirman <megous@megous.com>
Subject: [PATCH 4/4] bus: sunxi-rsb: Implement runtime power management
Date: Sun,  3 Jan 2021 05:06:35 -0600
Message-Id: <20210103110635.34823-5-samuel@sholland.org>
X-Mailer: git-send-email 2.26.2
In-Reply-To: <20210103110635.34823-1-samuel@sholland.org>
References: <20210103110635.34823-1-samuel@sholland.org>
MIME-Version: 1.0
X-CRM114-Version: 20100106-BlameMichelson ( TRE 0.8.0 (BSD) ) MR-646709E3 
X-CRM114-CacheID: sfid-20210103_060645_944475_F24BD852 
X-CRM114-Status: GOOD (  14.08  )
X-BeenThere: linux-arm-kernel@lists.infradead.org
X-Mailman-Version: 2.1.29
Precedence: list
List-Id: <linux-arm-kernel.lists.infradead.org>
List-Unsubscribe: 
 <http://lists.infradead.org/mailman/options/linux-arm-kernel>,
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=unsubscribe>
List-Archive: <http://lists.infradead.org/pipermail/linux-arm-kernel/>
List-Post: <mailto:linux-arm-kernel@lists.infradead.org>
List-Help: <mailto:linux-arm-kernel-request@lists.infradead.org?subject=help>
List-Subscribe: 
 <http://lists.infradead.org/mailman/listinfo/linux-arm-kernel>,
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=subscribe>
Cc: linux-sunxi@googlegroups.com, linux-kernel@vger.kernel.org,
 linux-arm-kernel@lists.infradead.org, Samuel Holland <samuel@sholland.org>
Sender: "linux-arm-kernel" <linux-arm-kernel-bounces@lists.infradead.org>
Errors-To: 
 linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@lists.infradead.org

Gate the clock to save power while the controller is idle.

Signed-off-by: Samuel Holland <samuel@sholland.org>
---
 drivers/bus/sunxi-rsb.c | 44 +++++++++++++++++++++++++++++++++++++++++
 1 file changed, 44 insertions(+)

diff --git a/drivers/bus/sunxi-rsb.c b/drivers/bus/sunxi-rsb.c
index efd222f36cdc..ba5100dfc413 100644
--- a/drivers/bus/sunxi-rsb.c
+++ b/drivers/bus/sunxi-rsb.c
@@ -46,6 +46,7 @@
 #include <linux/of_platform.h>
 #include <linux/platform_device.h>
 #include <linux/pm.h>
+#include <linux/pm_runtime.h>
 #include <linux/regmap.h>
 #include <linux/reset.h>
 #include <linux/slab.h>
@@ -337,6 +338,10 @@ static int sunxi_rsb_read(struct sunxi_rsb *rsb, u8 rtaddr, u8 addr,
 		return -EINVAL;
 	}
 
+	ret = pm_runtime_resume_and_get(rsb->dev);
+	if (ret)
+		return ret;
+
 	mutex_lock(&rsb->lock);
 
 	writel(addr, rsb->regs + RSB_ADDR);
@@ -352,6 +357,9 @@ static int sunxi_rsb_read(struct sunxi_rsb *rsb, u8 rtaddr, u8 addr,
 unlock:
 	mutex_unlock(&rsb->lock);
 
+	pm_runtime_mark_last_busy(rsb->dev);
+	pm_runtime_put_autosuspend(rsb->dev);
+
 	return ret;
 }
 
@@ -379,6 +387,10 @@ static int sunxi_rsb_write(struct sunxi_rsb *rsb, u8 rtaddr, u8 addr,
 		return -EINVAL;
 	}
 
+	ret = pm_runtime_resume_and_get(rsb->dev);
+	if (ret)
+		return ret;
+
 	mutex_lock(&rsb->lock);
 
 	writel(addr, rsb->regs + RSB_ADDR);
@@ -389,6 +401,9 @@ static int sunxi_rsb_write(struct sunxi_rsb *rsb, u8 rtaddr, u8 addr,
 
 	mutex_unlock(&rsb->lock);
 
+	pm_runtime_mark_last_busy(rsb->dev);
+	pm_runtime_put_autosuspend(rsb->dev);
+
 	return ret;
 }
 
@@ -672,10 +687,29 @@ static int sunxi_rsb_hw_init(struct sunxi_rsb *rsb)
 
 static void sunxi_rsb_hw_exit(struct sunxi_rsb *rsb)
 {
+	/* Keep the clock and PM reference counts consistent. */
+	if (pm_runtime_status_suspended(rsb->dev))
+		pm_runtime_resume(rsb->dev);
 	reset_control_assert(rsb->rstc);
 	clk_disable_unprepare(rsb->clk);
 }
 
+static int __maybe_unused sunxi_rsb_runtime_suspend(struct device *dev)
+{
+	struct sunxi_rsb *rsb = dev_get_drvdata(dev);
+
+	clk_disable_unprepare(rsb->clk);
+
+	return 0;
+}
+
+static int __maybe_unused sunxi_rsb_runtime_resume(struct device *dev)
+{
+	struct sunxi_rsb *rsb = dev_get_drvdata(dev);
+
+	return clk_prepare_enable(rsb->clk);
+}
+
 static int __maybe_unused sunxi_rsb_suspend(struct device *dev)
 {
 	struct sunxi_rsb *rsb = dev_get_drvdata(dev);
@@ -758,6 +792,12 @@ static int sunxi_rsb_probe(struct platform_device *pdev)
 	if (ret)
 		dev_warn(dev, "Initialize device mode failed: %d\n", ret);
 
+	pm_suspend_ignore_children(dev, true);
+	pm_runtime_set_active(dev);
+	pm_runtime_set_autosuspend_delay(dev, MSEC_PER_SEC);
+	pm_runtime_use_autosuspend(dev);
+	pm_runtime_enable(dev);
+
 	of_rsb_register_devices(rsb);
 
 	return 0;
@@ -768,6 +808,7 @@ static int sunxi_rsb_remove(struct platform_device *pdev)
 	struct sunxi_rsb *rsb = platform_get_drvdata(pdev);
 
 	device_for_each_child(rsb->dev, NULL, sunxi_rsb_remove_devices);
+	pm_runtime_disable(&pdev->dev);
 	sunxi_rsb_hw_exit(rsb);
 
 	return 0;
@@ -777,10 +818,13 @@ static void sunxi_rsb_shutdown(struct platform_device *pdev)
 {
 	struct sunxi_rsb *rsb = platform_get_drvdata(pdev);
 
+	pm_runtime_disable(&pdev->dev);
 	sunxi_rsb_hw_exit(rsb);
 }
 
 static const struct dev_pm_ops sunxi_rsb_dev_pm_ops = {
+	SET_RUNTIME_PM_OPS(sunxi_rsb_runtime_suspend,
+			   sunxi_rsb_runtime_resume, NULL)
 	SET_NOIRQ_SYSTEM_SLEEP_PM_OPS(sunxi_rsb_suspend, sunxi_rsb_resume)
 };
 
